[Bug]: cannot create network ... : conflicts with network ... : networks have overlapping IPv4

### Module

Core

### Testcontainers version

1.19.1

### Using the latest Testcontainers version?

No

### Host OS

Linux

### Host Arch

x86, ARM

### Docker version

```shell
Client: Docker Engine - Community
 Version:           27.0.2
 API version:       1.46
 Go version:        go1.21.11
 Git commit:        912c1dd
 Built:             Wed Jun 26 18:47:28 2024
 OS/Arch:           linux/amd64
 Context:           default
```


### What happened?

All of a sudden many of our GitHub action workflows started failing, because our tests couldn't start the containers. The error we see in the logs is:

```
Caused by: com.github.dockerjava.api.exception.DockerException: Status 403: ***"message":"cannot create network bda9d3d33ed5fdb18b1a5ffc80496e325e7e454f9292f05026cf8796b9aba395 (br-bda9d3d33ed5): conflicts with network bf05f12899f2e8b3b8835115218e4fa5074476c9d53676a31b7ace186d368b2a (br-bf05f12899f2): networks have overlapping IPv4"***
```

### Relevant log output

_No response_

### Additional Information

This has started only after Ubuntu published the new docker version 27 to its repository and our runners upgraded automatically, because we didn't pin the version. I see in the [release notes](https://docs.docker.com/engine/release-notes/27.0/) that there have been some changes to networking (and some networking APIs), so I'm wondering if this introduced some race condition.

In our tests, we create many networks concurrently using `Network.newNetwork()`, and that worked fine so far, as the docker daemon should be assigning subnets in an incrementing fashion.

I was wondering if this is maybe a regression in docker itself, and it's possible, but I wasn't able to reproduce locally with `for n in $(seq 0 256); do docker network create "n$n"& done.

--- a/libnetwork/ipams/defaultipam/address_space.go
+++ b/libnetwork/ipams/defaultipam/address_space.go
@@ -164,8 +164,6 @@ func (aSpace *addrSpace) allocatePredefinedPool(reserved []netip.Prefix) (netip.
 		pdf := aSpace.predefined[pdfID]
 
 		if allocated.Overlaps(pdf.Base) {
-			it.Inc()
-
 			if allocated.Bits() <= pdf.Base.Bits() {
 				// The current 'allocated' prefix is bigger than the 'pdf'
 				// network, thus the block is fully overlapped.
@@ -194,6 +192,8 @@ func (aSpace *addrSpace) allocatePredefinedPool(reserved []netip.Prefix) (netip.
 				// 'prevAlloc' and before 'allocated'.
 				return makeAlloc(afterPrev), nil
 			}
+
+			it.Inc()
 
 			if netiputil.LastAddr(allocated) == netiputil.LastAddr(pdf.Base) {
 				// The last address of the current 'allocated' prefix is the

--- a/libnetwork/ipams/defaultipam/address_space_test.go
+++ b/libnetwork/ipams/defaultipam/address_space_test.go
@@ -1,6 +1,7 @@
 package defaultipam
 
 import (
+	"fmt"
 	"net/netip"
 	"testing"
 
@@ -386,4 +387,114 @@ func TestStaticAllocation(t *testing.T) {
 		netip.MustParsePrefix("192.168.0.0/16"),
 		netip.MustParsePrefix("192.168.3.0/24"),
 	}, cmpopts.EquateComparable(netip.Prefix{})))
+}
+
+// Regression test for https://github.com/moby/moby/issues/48069
+func TestPoolAllocateAndRelease(t *testing.T) {
+	type testClosures struct {
+		alloc   func(netname string)
+		release func(netname string)
+	}
+
+	testcases := []struct {
+		name       string
+		predefined []*ipamutils.NetworkToSplit
+		reserved   []netip.Prefix
+		calls      []func(tcs testClosures)
+		tcs        testClosures
+	}{
+		{
+			name: "allocate after reserved",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("10.0.0.0/24"), Size: 24},
+				{Base: netip.MustParsePrefix("10.0.1.0/24"), Size: 24},
+				{Base: netip.MustParsePrefix("10.1.0.0/16"), Size: 24},
+			},
+			reserved: []netip.Prefix{
+				netip.MustParsePrefix("10.0.0.0/16"),
+			},
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc("n1") },
+				func(tcs testClosures) { tcs.alloc("n2") },
+			},
+		},
+		{
+			name: "reallocate first subnet",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("10.0.0.0/8"), Size: 24},
+			},
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc("n1") },
+				func(tcs testClosures) { tcs.alloc("n2") },
+				func(tcs testClosures) { tcs.alloc("n3") },
+				func(tcs testClosures) { tcs.release("n1") },
+				func(tcs testClosures) { tcs.alloc("n4") },
+				func(tcs testClosures) { tcs.alloc("n5") },
+			},
+		},
+		{
+			name: "reallocate after release",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("10.0.0.0/8"), Size: 24},
+			},
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc("n1") },
+				func(tcs testClosures) { tcs.alloc("n2") },
+				func(tcs testClosures) { tcs.alloc("n3") },
+				func(tcs testClosures) { tcs.alloc("n4") },
+				func(tcs testClosures) { tcs.release("n2") },
+				func(tcs testClosures) { tcs.release("n3") },
+				func(tcs testClosures) { tcs.alloc("n5") },
+				func(tcs testClosures) { tcs.alloc("n6") },
+				func(tcs testClosures) { tcs.alloc("n7") },
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			as, err := newAddrSpace(tc.predefined)
+			assert.NilError(t, err)
+			subnetToNetname := map[netip.Prefix]string{}
+			netnameToSubnet := map[string]netip.Prefix{}
+
+			// To avoid passing as,subnetToNetname,netnameToSubnet into each of the
+			// functions in tc.calls (cluttering the list of testcases), create closures
+			// that use them and pass those.
+			tcs := testClosures{
+				// Allocate a pool for netname, check that a subnet is returned that
+				// isn't already allocated, and doesn't overlap with a reserved range.
+				alloc: func(netname string) {
+					subnet, err := as.allocatePredefinedPool(tc.reserved)
+					assert.NilError(t, err)
+
+					otherNetname, exists := subnetToNetname[subnet]
+					assert.Assert(t, !exists, fmt.Sprintf(
+						"subnet %s allocated to %s, reallocated for %s", subnet, otherNetname, netname))
+					for _, reserved := range tc.reserved {
+						assert.Assert(t, !reserved.Overlaps(subnet),
+							fmt.Sprintf("subnet %s allocated in reserved range %s", subnet, reserved))
+					}
+
+					subnetToNetname[subnet] = netname
+					netnameToSubnet[netname] = subnet
+				},
+				// Release a pool for a netname - the test must ensure netname currently
+				// has an allocated subnet.
+				release: func(netname string) {
+					subnet, ok := netnameToSubnet[netname]
+					assert.Assert(t, ok)
+					err := as.releaseSubnet(subnet, netip.Prefix{})
+					assert.NilError(t, err)
+
+					delete(netnameToSubnet, netname)
+					delete(subnetToNetname, subnet)
+				},
+			}
+
+			for _, f := range tc.calls {
+				f(tcs)
+			}
+		})
+	}
 }

